### ISSUE TITLE
Fx for iOS - Links in the Terms of Service or Privacy Policy replace FxA

### ISSUE THREAD

**Issue by shane-tomlinson** [mozilla/fxa-content-server#2225]
STR

1. Install Fx for iOS.
1. Open FxA from Settings
1. Open the "Privacy Notice"
1. Click in the "Mozilla Privacy Policy" link.
1. Where did FxA go?

ref #1968, #2044, #2326

**Comment by shane-tomlinson** [comment]
cc @ncalexan

**Comment by shane-tomlinson** [comment]
ref #1968 - a similar problem manifests in the iframe flow. 

Proposed solution - if in the iframe or iOS, open links from the TOS/PP agreements in new tabs.

**Comment by shane-tomlinson** [comment]
@ncalexan - if we add a `target="_blank"` attribute to the TOS/PP links, will the pages open in a new tab? Is there an equivalent to about:config that I can use to set Fennec on iOS to point to a local FxA install to test?

**Comment by shane-tomlinson** [comment]
> @ncalexan - if we add a target="_blank" attribute to the TOS/PP links, will the pages open in a new tab? Is there an equivalent to about:config that I can use to set Fennec on iOS to point to a local FxA install to test?

You know, we have one link with a `target="_blank"` attribute, and we have a bug filed that the link does nothing (#2226). Adding `target="_blank"` will not work for iOS.

**Comment by shane-tomlinson** [comment]
@johngruen and @ryanfeeley - what do you think of @vladikoff's proposal and placing the links next to the text in brackets like so:

Privacy

![screen shot 2015-04-15 at 17 49 55](https://cloud.githubusercontent.com/assets/848085/7164428/0f477d42-e398-11e4-914d-30a06fbe14aa.png)

Terms

![screen shot 2015-04-15 at 17 50 09](https://cloud.githubusercontent.com/assets/848085/7164431/11c85e88-e398-11e4-9452-1689b061dc6d.png)

**Comment by johngruen** [comment]
@shane-Tomlinson: lgtm

**Comment by ryanfeeley** [comment]
Probably better from a legal perspective too? Should we inquire?

**Comment by johngruen** [comment]
@ryanfeeley: no

**Comment by johngruen** [comment]
Seriously, there's very little legal is going to add here, and this is a commonly used pattern on other sites.

**Comment by pdehaan** [comment]
@shane-tomlinson Looks great!
Not sure how much logic we can add since it's only CSS, or if we can not do the ::after href content if the link is already a URL.
For example, that bit in your screenshot where it says:

> "... you can log into https://find.firefox.com [https://find.firefox.com] using your ..."

**Comment by pdehaan** [comment]
Also, not sure if we want to add any styling to the non-link links. For example, adding an <u>underline</u> like in @vladikoff's https://github.com/mozilla/fxa-content-server/issues/1968#issuecomment-91246231, or maybe make the text **bold** so it's more obvious exactly what text is associated with the specified URL.

**Comment by ncalexan** [comment]
On Thu, Apr 9, 2015 at 5:48 AM, Shane Tomlinson <notifications@github.com>
wrote:

> @ncalexan <https://github.com/ncalexan> - if we add a target="_blank"
> attribute to the TOS/PP links, will the pages open in a new tab? Is there
> an equivalent to about:config that I can use to set Fennec on iOS to point
> to a local FxA install to test?
>
> You know, we have one link with a target="_blank" attribute, and we have
> a bug filed that the link does nothing (#2226
> <https://github.com/mozilla/fxa-content-server/issues/2226>). Adding
> target="_blank" will not work for iOS.
>
It's probable we could make `target="_blank"` work on the iOS webview,
perhaps opening a new regular tab.  However, the navigation back would be
quite challenging to arrange, so I'd prefer not to do it.

**Comment by shane-tomlinson** [comment]
Thanks for all the research @pdehaan. I'm not terribly worried about the `find.firefox.com` link, and I was hoping to avoid JavaScript. With #1968 we'll probably have to add JS to add `target="_blank"` attributes to the links anyways, so it could be nice to clean this up.

Underlining the text associated with the link is a great idea.

**Comment by shane-tomlinson** [comment]
@johngruen and @ryanfeeley - I have a new proposal, taking into consideration #2326, #2044, #1968.

For Sync in Fx on iOS - disable the link, show the href next to the original text as shown above.
For everyone else (including the iframe) - once the template is rendered, add a `target="_blank"` attribute, open the links in an external page.

This allows Sync in Fx on iOS users to copy/paste the link if needed. It allows users on other browsers to open the links and not lose the back button.

**Comment by shane-tomlinson** [comment]
> For everyone else (including the iframe) - once the template is rendered, add a target="_blank" attribute, open the links in an external page.

I have a local fix for the back button disappearing (#2044) after browsing away or page refresh, so target="_blank" would only need to be added if the user was in an iframe.

**Comment by shane-tomlinson** [comment]
This also happens from Sync:

![screen shot 2015-04-16 at 19 05 58](https://cloud.githubusercontent.com/assets/848085/7187883/c202ddf4-e46b-11e4-9e18-62fe010ebf74.png)

**Comment by pdehaan** [comment]
I was playing with this last night, and thought I'd throw the code here since it's marginally interesting...
If we decide we _DO_ want to fix the edge case of the duplicated "https://find.firefox.com", we could use something like [**grunt-md**](https://www.npmjs.com/package/grunt-md) and make use of the `preCompile` and `postCompile` hooks (which I think come from the excellently named [**marky-mark**](https://www.npmjs.com/package/marky-mark) Markdown parser).

```js
'use strict';

var cheerio = require('cheerio');

module.exports = function (grunt) {
  grunt.loadNpmTasks('grunt-md');

  grunt.initConfig({
    md: {
      options: {
        mm: {
          preCompile: function (md) {
            return md.replace(/{:\s.*?\s}/g, '');
          },
          postCompile: function (html) {
            var $ = cheerio.load(html);
            $('a').each(function (i, elem) {
              var caption = $(elem).text();
              // elem.attribs['data-caption'] = caption;
              elem.attribs['data-is-url-caption'] = isHref(caption).toString();
            });
            return $.html();
          }
        }
      },
      app: {
        src: 'bower_components/legal-docs/firefox_cloud_services_{PrivacyNotice,ToS}/*.md',
        dest: 'out'
      }
    }
  });
};

function isHref(text) {
  return (/^https?\:/i).test(text);
}
```

Basically that code uses the `preCompile` hook to remove the super weird `{: datetime="2014-08-29" }` metadata before parsing the markdown, and then uses the `postCompile` hook to use **cheerio** and it's awesome jQuery-like parsing to loop over all `<a>` tags and add a Boolean `data-is-url-caption` attribute if the link's inner `text()` starts with "http:" or "https:".
Now we'd be able to only use the :after and `content` trick if the `data-is-url-caption` attribute is false:
```css
  a[data-is-url-caption="false"]:after {
    color: red;
    content: " [" attr(href) "]";
  }
```

![nightly](https://cloud.githubusercontent.com/assets/557895/7189152/baafe5e0-e431-11e4-97bf-a4be4d672f59.png)

---

**TL;DR:** The **grunt-md** and **cheerio** packages are pretty cool!

**Comment by shane-tomlinson** [comment]
@pdehaan - I took your idea and did it in the view after the template was inserted into the DOM but before it's made visible:

```js
        self.$('a').each(function (index, element) {
          element = $(element);
          var href = element.attr('href');

          // data-url is used for frames and Sync in FxiOS to display
          // the href in brackets next to the original text.
          // Only show the href if the href is different from
          // the text. See #2225
          if (element.attr('href') !== element.text()) {
            element.attr('data-url', href);
          }
        });
```

The corresponding CSS:
```css
  // Links cannot be opened from the TOS/PP text when signing
  // in to Sync on Fx for iOS. Show the hrefs next to the link.
  // The href is fetched from the data-url attribute instead of
  // the href attribute because some links are the same as their
  // text. In those cases, do not show both the original link and
  // the converted link, just show the original link.
  .fx-ios-sync &, .iframe & {
    a[href] {
      // using text-decoration: underline underlines the ::after
      // section as well, with no way to remove it.
      // So, add a border to the entire element, then hide
      // the border in the ::after using a border that is the same
      // color as the background.
      border-bottom: 1px dotted $text-color;
      color: $text-color;
      cursor: default;
      pointer-events: none;
      text-decoration: none;
    }

    a[data-url^=http]::after {
      border-bottom: 1px solid $content-background-color;
      content: " [" attr(data-url) "] ";
    }
  }
```
We don't really care if users who browse directly to the TOS/PP pages can browse away, but we do if they open the TOS/PP in Sync for FxiOS (or maybe any Sync for that matter), or in an iframe.

**Comment by shane-tomlinson** [comment]
Here's the result:

![screen shot 2015-04-16 at 20 44 34](https://cloud.githubusercontent.com/assets/848085/7189826/8190011c-e479-11e4-9f59-cea5e9fd0de9.png)

**Comment by pdehaan** [comment]
@shane-tomlinson:

![](http://38.media.tumblr.com/f22cd544bdbc7d8a7004445242781ed0/tumblr_mud534aJvq1stmkupo1_400.gif)
